jamtis changes

jeffro256 opened the pull request:
* nominal address tag protection for LWs (extra address key)
* flexible view tags
* churning and pocketchange protection for LWs (auxiliary enote records)

jeffro256 committed:
jamtis changes

* nominal address tag protection for LWs (extra address key)
* flexible view tags
* churning and pocketchange protection for LWs (auxiliary enote records)

UkoeHB committed:
seraphis mega squash commit [initial commit - 05/15/2023]; see seraphis_lib_hist_05_15_23 branch for commit history

UkoeHB committed:
add key exchange round booster to multisig_account

UkoeHB committed:
async: explicit task constructors

UkoeHB committed:
rebase fixes

DangerousFreedom1984 committed:
make JamtisDestinationV1 serializable (#11)

make JamtisDestinationV1 serializable 

---------

Co-authored-by: DangerousFreedom <monero-inflation-checker@protonmail.com>

UkoeHB committed:
compile fix

UkoeHB committed:
clarify SpBasicEnoteRecordV1 docs; add async unit test for custom joins

UkoeHB committed:
remove broken perf tests

DangerousFreedom1984 committed:
add operator== to JamtisPaymentProposals (#18)

* add operator== to JamtisPaymentProposals

---------

Co-authored-by: DangerousFreedom <monero-inflation-checker@protonmail.com>\

DangerousFreedom1984 committed:
make JamtisPaymentProposal serializable (#21)

* make JamtisPaymentProposal serializable

---------

Co-authored-by: DangerousFreedom <monero-inflation-checker@protonmail.com>

j-berman committed:
util to convert legacy outputs to seraphis lib compatible enotes (#22)

Commit:
container_helpers: is_sorted_and_unique 3x fewer comparison ops (#19)

DangerousFreedom1984 committed:
modify construct_tx_for_mock_ledger_v1 so it outputs the JamtisPaymentProposals (#24)

* modify construct_tx_for_mock_ledger_v1 so it outputs the JamtisPaymentProposals

---------

Co-authored-by: DangerousFreedom <monero-inflation-checker@protonmail.com>

jeffro256 commented:
For unmodified `seraphis_lib`, running the `test_remote_scanner_client_scan_sp` performance test, I got the following result:

```
test_remote_scanner_client_scan_sp (23 calls) - OK: 13391 µs/call
test_remote_scanner_client_scan_sp (23 calls) - OK: 13478 µs/call
test_remote_scanner_client_scan_sp (23 calls) - OK: 14521 µs/call
```

On this branch, I get:

```
test_remote_scanner_client_scan_sp (23 calls) - OK: 1790 ms/call
test_remote_scanner_client_scan_sp (23 calls) - OK: 1811 ms/call
test_remote_scanner_client_scan_sp (23 calls) - OK: 1802 ms/call
```

This means that for the different test modes `ALL_FAKE`, `ONE_FAKE_TAG_MATCH`, `ONE_OWNED`, the new client-side view scanning takes 134x, 134x, and 124x, respectively, times more CPU cycles than unmodified `seraphis_lib`. This puts it on par with the dense/sparse changes #15.

jeffro256 commented:
Latest commit adds a unit test case which performs a successful Janus attack with knowledge of one address private key. Will push fix soon...

jeffro256 commented:
Currently making modifications to the "Implementing Seraphis" document to reflect the changes in this PR. Will release a PR to the Seraphis repo when done.

jeffro256 commented:
Jamtis changes are documented here: https://github.com/UkoeHB/Seraphis/pull/6

UkoeHB committed:
rebase fixes

UkoeHB committed:
fix unbound issue

UkoeHB committed:
missing header

UkoeHB committed:
c++17

UkoeHB committed:
c++17

jeffro256 commented:
Rebased

Commit:
split variant into variant and optional (#28)

Commit:
unit_tests: variant: fix for optional_variant (#30)

UkoeHB commented:
Needs to be updated to reflect the notational changes made in https://github.com/UkoeHB/Seraphis/pull/6.

DangerousFreedom1984 committed:
derive view_balance from master key (#29)

* derive view_balance from master key

Co-authored-by: UkoeHB <37489173+UkoeHB@users.noreply.github.com>

---------

Co-authored-by: DangerousFreedom1984 <monero-inflation-checker@protonmail.com>
Co-authored-by: UkoeHB <37489173+UkoeHB@users.noreply.github.com>

UkoeHB committed:
fix typo in header

jeffro256 commented:
Did a simple rebase again to fix merge conflicts with the derived view balance key PR. Will tackle updating notation today

jeffro256 commented:
How does the notation look on first glance?

UkoeHB commented on src/crypto/x25519.h:
```diff
@@ -121,3 +121,7 @@ void x25519_invmul_key(std::vector<x25519_secret_key> privkeys_to_invert,
+
+inline const unsigned char* to_bytes(const crypto::x25519_pubkey &point) { return &reinterpret_cast<const unsigned char&>(point); }
+inline const unsigned char* to_bytes(const crypto::x25519_scalar &scalar) { return &reinterpret_cast<const unsigned char&>(scalar); }
+inline const unsigned char* to_bytes(const crypto::x25519_secret_key &skey) { return &reinterpret_cast<const unsigned char&>(skey); }
```
These should go above the `_MAKE_HASHABLE` to be consistent with `crypto.cpp`.

UkoeHB commented on src/seraphis_core/jamtis_account_secrets.h:
```diff
@@ -60,47 +60,56 @@ namespace jamtis
+ * brief make_jamtis_exchangebase_pubkey - D_base
+ * D_base = d_vr * xG
+ * param: d_view_received - d_vr
+ * outparam: baseexchange_pubkey_out
```
```suggestion
 * outparam: exchangebase_pubkey_out
```

UkoeHB commented on src/seraphis_core/jamtis_account_secrets.h:
```diff
@@ -60,47 +60,56 @@ namespace jamtis
-*   xk_fr = H_n_x25519[k_vb]()
-* param: k_view_balance - k_vb
-* outparam: xk_find_received_out - xk_fr
+* brief: make_jamtis_filterassist_key - filter-assist key, for calculating primary view tag
```
```suggestion
* brief: make_jamtis_filterassist_key - filter-assist key, for calculating primary view tags
```

UkoeHB commented on src/seraphis_core/jamtis_account_secrets.h:
```diff
@@ -60,47 +60,56 @@ namespace jamtis
-* outparam: findreceived_pubkey_out - xK_fr
+* brief: make_jamtis_filterassist_pubkey - D_fa
+*   - D_fa = d_fa * D_base
+* param: d_filter_asiist - d_fa
```
```suggestion
* param: d_filter_assist - d_fa
```

UkoeHB commented on src/seraphis_core/jamtis_address_tag_utils.cpp:
```diff
@@ -87,48 +87,8 @@ static encrypted_address_tag_secret_t get_encrypted_address_tag_secret(const rct
-    hash_context.cipher_key = rct::sk2rct(cipher_key);
-    hash_context.enc_j = encrypted_address_index;
-
-    // address_tag_hint = H_2(k, cipher[k](j))
-    address_tag_hint_t address_tag_hint;
-    sp_hash_to_2(&hash_context, sizeof(hash_context), address_tag_hint.bytes);
-
```
Sad to see my genius address tag hint going away lol.

jeffro256 replied:
RIP. It was a really elegant solution while it lasted

UkoeHB commented on src/seraphis_core/jamtis_address_tag_utils.h:
```diff
@@ -67,23 +67,22 @@ struct jamtis_address_tag_cipher_context final
 address_tag_t cipher_address_index(const jamtis_address_tag_cipher_context &cipher_context, const address_index_t &j);
 address_tag_t cipher_address_index(const crypto::secret_key &cipher_key, const address_index_t &j);
 
 /// try to get j from an address tag
```
```suggestion
/// get j from an address tag
```

UkoeHB commented on src/seraphis_core/jamtis_destination.h:
```diff
@@ -54,12 +54,14 @@ namespace jamtis
-    /// xK_3 = xk^j_a xK_ua                       (DH base key)
-    crypto::x25519_pubkey addr_K3;
+    /// K^j_s     = k^j_g G + k^j_x X + k^j_u U + K_s (address spend key)
+    rct::key addr_s;
```
```suggestion
    rct::key addr_Ks;
```

Lowercase letter implies private key or scalar.

UkoeHB commented on src/seraphis_core/jamtis_destination.cpp:
```diff
@@ -54,42 +54,50 @@ namespace jamtis
+    make_jamtis_address_privkey(spend_pubkey, s_generate_address, j, address_privkey);
 
-    x25519_scmul_key(address_privkey, findreceived_pubkey, destination_out.addr_K2);
+    /// D^j_fa = d^j_a * D_fa
```
```suggestion
    // D^j_fa = d^j_a * D_fa
```

UkoeHB commented on src/seraphis_core/jamtis_destination.cpp:
```diff
@@ -54,42 +54,50 @@ namespace jamtis
 
-    // xK_3 = xk^j_a xK_ua
-    x25519_scmul_key(address_privkey, unlockamounts_pubkey, destination_out.addr_K3);
+    /// D^j_vr = d^j_a * D_vr
```
```suggestion
    // D^j_vr = d^j_a * D_vr
```

UkoeHB commented on src/seraphis_core/jamtis_destination.cpp:
```diff
@@ -54,42 +54,50 @@ namespace jamtis
+    x25519_scmul_key(address_privkey, viewreceived_pubkey, destination_out.addr_Dvr);
 
-    // addr_tag = cipher[k](j) || H_2(k, cipher[k](j))
+    /// D^j_base = d^j_a * D_base
```
```suggestion
    // D^j_base = d^j_a * D_base
```

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.h:
```diff
@@ -49,35 +49,37 @@ namespace jamtis
 * outparam: view_tag_out - view_tag
 */
-void make_jamtis_view_tag(const crypto::x25519_pubkey &sender_receiver_DH_derivation,
+void make_jamtis_flagship_view_tag(const crypto::x25519_pubkey &dhe_fa,
```
```suggestion
void make_jamtis_standard_view_tag(const crypto::x25519_pubkey &dhe_fa,
```

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.h:
```diff
@@ -269,22 +272,51 @@ bool test_jamtis_onetime_address(const rct::key &recipient_address_spend_key,
-* param: enote_ephemeral_pubkey - xK_e
-* param: input_context - input_context
+* brief: test_jamtis_primary_view_tag - test primary view tag
+* param: dhe_fa - D^d_fa = [sender: xr * D^j_af] = [recipient: d_fa * D_e]
```
```suggestion
* param: dhe_fa - D^d_fa = [sender: xr * D^j_fa] = [recipient: d_fa * D_e]
```

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -153,39 +153,93 @@ static void make_jamtis_amount_baked_key_plain(const crypto::x25519_pubkey &reve
 }
 //-------------------------------------------------------------------------------------------------------------------
 //-------------------------------------------------------------------------------------------------------------------
+static inline std::uint32_t vttou32(view_tag_t vt)
```
At least add a doc comment to this opaquely named function...

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -153,39 +153,93 @@ static void make_jamtis_amount_baked_key_plain(const crypto::x25519_pubkey &reve
+    const rct::key &onetime_address,
+    view_tag_t &naked_primary_view_tag_out)
+{
+    static_assert(2 * VIEW_TAG_BYTES <= 8, "need to expand hash function");
```
```suggestion
    static_assert(VIEW_TAG_BYTES <= 8, "need to expand hash function");
```

jeffro256 replied:
Sorry, this is a holdover from the primary view tag residue concept

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -153,39 +153,93 @@ static void make_jamtis_amount_baked_key_plain(const crypto::x25519_pubkey &reve
+{
+    static_assert(VIEW_TAG_BYTES == 2, "sp_hash_to_2/VIEW_TAG_BYTES output mismatch");
+
+    // H_ncbits(q)
```
```suggestion
    // H_2(q)
```

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -153,39 +153,93 @@ static void make_jamtis_amount_baked_key_plain(const crypto::x25519_pubkey &reve
+    static_assert(VIEW_TAG_BYTES == 2, "sp_hash_to_2/VIEW_TAG_BYTES output mismatch");
+
+    // H_ncbits(q)
+    SpKDFTranscript transcript2{config::HASH_KEY_JAMTIS_VIEW_TAG_PRIMARY, sizeof(rct::key)};
```
```suggestion
    SpKDFTranscript transcript2{config::HASH_KEY_JAMTIS_VIEW_TAG_COMPLEMENTARY, sizeof(rct::key)};
```

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -153,39 +153,93 @@ static void make_jamtis_amount_baked_key_plain(const crypto::x25519_pubkey &reve
 }
 //-------------------------------------------------------------------------------------------------------------------
-void make_jamtis_view_tag(const crypto::x25519_pubkey &sender_receiver_DH_derivation,
+void make_jamtis_flagship_view_tag(const crypto::x25519_pubkey &dhe_fa,
```
```suggestion
void make_jamtis_standard_view_tag(const crypto::x25519_pubkey &dhe_fa,
```

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -153,39 +153,93 @@ static void make_jamtis_amount_baked_key_plain(const crypto::x25519_pubkey &reve
-    SpKDFTranscript transcript{config::HASH_KEY_JAMTIS_VIEW_TAG, 2*sizeof(rct::key)};
-    transcript.append("xK_d", sender_receiver_DH_derivation);
-    transcript.append("Ko", onetime_address);
+    // naked_primary_view_tag = H_npbits(D^d_fa, Ko)
```
```suggestion
    // naked_primary_view_tag = H_8(D^d_fa, Ko)
```

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -153,39 +153,93 @@ static void make_jamtis_amount_baked_key_plain(const crypto::x25519_pubkey &reve
+    view_tag_t naked_primary_view_tag;
+    make_jamtis_naked_primary_view_tag(dhe_fa, onetime_address, naked_primary_view_tag);
+
+    // naked_complementary_view_tag = H_ncbits(q)
```
```suggestion
    // naked_complementary_view_tag = H_2(q)
```

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -153,39 +153,93 @@ static void make_jamtis_amount_baked_key_plain(const crypto::x25519_pubkey &reve
+    const std::uint32_t comp_mask = ~primary_mask;
 
-    sp_hash_to_1(transcript.data(), transcript.size(), &view_tag_out);
+    // view_tag = naked_primary_view_tag[:npbits] || naked_complementary_view_tag[:ncbits]
+    std::uint32_t combined_view_tag_u32 = (vttou32(naked_primary_view_tag) & primary_mask) |
+        ((vttou32(naked_complementary_view_tag) << num_primary_view_tag_bits) & comp_mask);
```
Pretty sure the `comp_mask` here is redundant, but redundancy is ok (and it's cheap in this case).

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -153,39 +153,93 @@ static void make_jamtis_amount_baked_key_plain(const crypto::x25519_pubkey &reve
+    std::uint32_t combined_view_tag_u32 = (vttou32(naked_primary_view_tag) & primary_mask) |
+        ((vttou32(naked_complementary_view_tag) << num_primary_view_tag_bits) & comp_mask);
+
+    combined_view_tag_u32 = SWAP32LE(combined_view_tag_u32);
+    memcpy(view_tag_out.bytes, &combined_view_tag_u32, VIEW_TAG_BYTES);
 }
```
I understand why you did this, but moving in and out of the environment's endianness is kinda awkward.

jeffro256 replied:
Should I hide that awkwardness behind a conversion function like `u32tovt()`?

UkoeHB replied:
No it's fine, just commenting for the sake of commenting.

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -253,12 +307,12 @@ void make_jamtis_sender_receiver_secret_plain(const crypto::x25519_secret_key &p
-    // q = H_32[k_vb](xK_e, input_context)
-    SpKDFTranscript transcript{selfsend_sender_receiver_secret_domain_separator(self_send_type), 2*sizeof(rct::key)};
-    transcript.append("xK_e", enote_ephemeral_pubkey);
+    // q = H_32[k_vb](is_aux, D_e, input_context)
```
```suggestion
    // q = H_32[k_vb](D_e, input_context)
```
`is_aux` isn't hashed, it's implicit in the domain separator.

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.h:
```diff
@@ -127,24 +129,24 @@ void make_jamtis_sender_receiver_secret_plain(const crypto::x25519_secret_key &p
 /**
 * brief: make_jamtis_sender_receiver_secret_selfsend - sender-receiver secret q for a self-send enote of a specific type
-*    q = H_32[k_vb](xK_e, input_context)
+*    q = H_32[k_vb](is_aux, D_e, input_context)
```
```suggestion
*    q = H_32[k_vb](D_e, input_context)
```

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -417,27 +470,73 @@ bool test_jamtis_onetime_address(const rct::key &recipient_address_spend_key,
-    // recompute view tag and check that it matches; short-circuit on failure
-    view_tag_t recomputed_view_tag;
-    make_jamtis_view_tag(sender_receiver_DH_derivation, onetime_address, recomputed_view_tag);
+    // npbits can't be greater than total tag size (duh)
+    if (num_primary_view_tag_bits > 8 * VIEW_TAG_BYTES)
+        return false;
```
Do an assert here, invalid bit length should be considered an upstream bug.

jeffro256 replied:
a throwable assert or just an assert that logs an error message?

UkoeHB replied:
Yeah `CHECK_AND_ASSERT_THROW_MSG`

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -417,27 +470,73 @@ bool test_jamtis_onetime_address(const rct::key &recipient_address_spend_key,
+    crypto::x25519_pubkey dhe_1;
+    crypto::x25519_scmul_key(d_filter_assist, enote_ephemeral_pubkey, dhe_1);
+
+    return test_jamtis_primary_view_tag(dhe_1,
```
```suggestion
    crypto::x25519_pubkey dhe_fa;
    crypto::x25519_scmul_key(d_filter_assist, enote_ephemeral_pubkey, dhe_fa);

    return test_jamtis_primary_view_tag(dhe_fa,
```

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -417,27 +470,73 @@ bool test_jamtis_onetime_address(const rct::key &recipient_address_spend_key,
+{
+    // npbits can't be greater than total tag size (duh)
+    if (num_primary_view_tag_bits > 8 * VIEW_TAG_BYTES)
         return false;
```
Assert here

UkoeHB reviewed (changes requested):
13/72 files done.

One thing you need to do is check that all the KDF transcripts are <= 128 bytes (i.e. one blake2b block). If > 128 then the cost to hash increases a lot. You can test this by editing `SpKDFTranscript` so the constructor prints the domain separator, and the `.size()` method prints its size. Then when you run the seraphis unit test suite all the sizes will print.

If you find any that crept over the line, then the relevant domain separator needs to be shaved down.

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -476,8 +472,8 @@ bool test_jamtis_primary_view_tag(const crypto::x25519_pubkey &dhe_fa,
     // npbits can't be greater than total tag size (duh)
-    if (num_primary_view_tag_bits > 8 * VIEW_TAG_BYTES)
-        return false;
+    CHECK_AND_ASSERT_THROW_MES(num_primary_view_tag_bits < 8 * VIEW_TAG_BYTES,
```
Needs to be <=.

jeffro256 commented on src/seraphis_main/scan_balance_recovery_utils.h:
```diff
@@ -101,7 +101,12 @@ bool try_find_legacy_enotes_in_tx(const rct::key &legacy_base_spend_pubkey,
+    const SpTxSupplementV1 &tx_supplement,
+    const std::vector<SpEnoteVariant> &enotes_in_tx,
+    std::vector<bool>& is_auxiliary_out);
+bool try_find_sp_enotes_in_tx(const crypto::x25519_secret_key &d_filter_assist,
```
The idea behind there being two `try_find_sp_enotes_in_tx` is that the first one does the meat of the work, but does the least amount of allocations and copying possible. All it really needs to allocate is the output vector of bools. This ideally would be used in actual production code which scans on behalf of users, but doesn't need to send out chunk data yet, and just caches some bools persistently (maybe as unsigned ints for compactness).

UkoeHB commented on tests/unit_tests/seraphis_knowledge_proofs.cpp:
```diff
@@ -566,18 +566,18 @@ TEST(seraphis_knowledge_proofs, sp_all_knowledge_proofs)
 
     // b. legacy user address
-    rct::key legacy_subaddr_spendkey_A;
+    rct::key legacy_subaddr_Kspendkey_A;
```
Search and replace mistake

UkoeHB commented on src/seraphis_core/jamtis_enote_utils.cpp:
```diff
@@ -153,39 +153,89 @@ static void make_jamtis_amount_baked_key_plain(const crypto::x25519_pubkey &reve
+
+    // H_2(D^d_fa, Ko)
+    unsigned char hash_temp[8];
+    SpKDFTranscript transcript1{config::HASH_KEY_JAMTIS_VIEW_TAG_PRIMARY, 2*sizeof(rct::key)};
```
```suggestion
    // H_2(D^d_fa, Ko)
    SpKDFTranscript transcript1{config::HASH_KEY_JAMTIS_VIEW_TAG_PRIMARY, 2*sizeof(rct::key)};
```

jeffro256 commented:
In regards to the KDF transcripts being smaller than 128 bytes, I applied this diff:
```
diff --git a/src/seraphis_crypto/sp_transcript.h b/src/seraphis_crypto/sp_transcript.h
index 17db163d6..f0e192fb3 100644
--- a/src/seraphis_crypto/sp_transcript.h
+++ b/src/seraphis_crypto/sp_transcript.h
@@ -100,7 +100,7 @@ public:
 
     /// access the transcript data
     const void* data() const { return m_transcript.data(); }
-    std::size_t size() const { return m_transcript.size(); }
+    std::size_t size() const { if (m_transcript.size() > 128) { std::cout << "BIG TRANSCRIPT: " << m_transcript.data(); } return m_transcript.size(); }
 
 private:
 //member variables
```

I then ran `./build/Linux/jamtis_lw_take_2/debug/tests/unit_tests/unit_tests --gtest_filter=seraphis* | grep jamtis_`. The only domain separator that showed up was `jamtis_input_context_standardlegacy_input_KI`, so I don't think that these changes should've made any transcripts go over 128 bytes.

Commit:
free function serialization DSL (#32)

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -115,30 +138,14 @@ bool operator==(const JamtisPaymentProposalSelfSendV1 a, const JamtisPaymentProp
-    // enote ephemeral pubkey: xK_e = xr xK_3
-    make_jamtis_enote_ephemeral_pubkey(proposal.enote_ephemeral_privkey,
-        proposal.destination.addr_K3,
+    get_enote_ephemeral_pubkey(proposal.enote_ephemeral_privkey, proposal.destination,
         enote_ephemeral_pubkey_out);
```
```suggestion
    get_enote_ephemeral_pubkey(proposal.enote_ephemeral_privkey, proposal.destination, enote_ephemeral_pubkey_out);
```

jeffro256 replied:
Why this change?

jeffro256 replied:
Nevermind, the indentation showed up weird for me on the browser. I understand

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -115,30 +138,14 @@ bool operator==(const JamtisPaymentProposalSelfSendV1 a, const JamtisPaymentProp
-    make_jamtis_enote_ephemeral_pubkey(proposal.enote_ephemeral_privkey,
-        proposal.destination.addr_K3,
+    get_enote_ephemeral_pubkey(proposal.enote_ephemeral_privkey, proposal.destination,
         enote_ephemeral_pubkey_out);
```
```suggestion
    get_enote_ephemeral_pubkey(proposal.enote_ephemeral_privkey, proposal.destination, enote_ephemeral_pubkey_out);
```

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -160,22 +167,29 @@ void get_coinbase_output_proposal_v1(const JamtisPaymentProposalV1 &proposal,
+    crypto::x25519_scmul_key(proposal.enote_ephemeral_privkey, proposal.destination.addr_Dfa, dhe_fa);
 
-    // 5. sender-receiver shared secret (plain): q = H_32(xK_d, xK_e, input_context)
+    // 4. derived key: D^d_vr = xr * D^j_vr
```
Comment numbering got messed up here

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -200,16 +214,20 @@ void get_output_proposal_v1(const JamtisPaymentProposalV1 &proposal,
+    crypto::x25519_pubkey dhe_fa; auto dhe1_wiper = auto_wiper(dhe_fa);
+    crypto::x25519_scmul_key(proposal.enote_ephemeral_privkey, proposal.destination.addr_Dfa, dhe_fa);
+
+    // 3. derived key: D^d_vr = xr * D^j_vr
```
Numbering also messed up here

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -81,21 +81,44 @@ static void get_output_proposal_amount_parts_v1(const rct::key &q,
+    const std::uint8_t num_primary_view_tag_bits,
     rct::key &onetime_address_out,
     encrypted_address_tag_t &addr_tag_enc_out,
     view_tag_t &view_tag_out)
```
Make two separate functions for aux/non-aux. Once you have parameters that are only needed by one or another branch, that's a good sign to make more functions.

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -258,24 +279,24 @@ void get_output_proposal_v1(const JamtisPaymentProposalSelfSendV1 &proposal,
         q);
 
-    // 5. amount baked key (selfsend): H_32[k_vb](q)
+    // 5. amount baked key (selfsend): H_32(self_send_type, q)
```
```suggestion
    // 5. amount baked key (selfsend): H_32[k_vb](q)
```
Domain separator stuff isn't expressed. Also, k_vb is used here.

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -299,13 +323,15 @@ void get_output_proposal_v1(const JamtisPaymentProposalSelfSendV1 &proposal,
     temp.destination             = gen_jamtis_destination_v1();
     temp.amount                  = amount;
     temp.enote_ephemeral_privkey = crypto::x25519_secret_key_gen();
+    temp.num_primary_view_tag_bits = num_primary_view_tag_bits;
```
Fix alignment

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -326,6 +352,7 @@ JamtisPaymentProposalSelfSendV1 gen_jamtis_selfsend_payment_proposal_v1(const rc
     temp.amount                  = amount;
     temp.type                    = type;
     temp.enote_ephemeral_privkey = crypto::x25519_secret_key_gen();
+    temp.num_primary_view_tag_bits = crypto::rand_idx<size_t>(8 * VIEW_TAG_BYTES);
```
Fix alignment

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.h:
```diff
@@ -133,7 +137,8 @@ void get_coinbase_output_proposal_v1(const JamtisPaymentProposalV1 &proposal,
 * outparam: enote_ephemeral_pubkey_out -
 * outparam: encoded_amount_out -
 * outparam: addr_tag_enc_out -
-* outparam: view_tag_out -
+* outparam: naked_primary_view_tag_out -
+* outparam: naked_complementary_view_tag_out -
```
Leftover?

jeffro256 replied:
Yes

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.h:
```diff
@@ -153,7 +158,8 @@ void get_output_proposal_v1(const JamtisPaymentProposalV1 &proposal,
-* outparam: view_tag_out -
+* outparam: naked_primary_view_tag_out -
+* outparam: naked_complementary_view_tag_out -
 * outparam: partial_memo_out -
```
Leftover?

UkoeHB commented on src/seraphis_core/jamtis_support_types.h:
```diff
@@ -26,7 +26,7 @@
 // THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-// Supporting types for Jamtis (address index, address tag hint, address tag, etc.).
+//! @file Supporting types for Jamtis (address index, address tag, double edged view tag, etc.).
```
```suggestion
//! @file Supporting types for Jamtis (address index, address tag, view tag, etc.).
```

UkoeHB commented on src/seraphis_core/jamtis_support_types.h:
```diff
@@ -57,50 +57,48 @@ struct address_index_t final
-    address_tag_hint_t();
-};
-
 /// index ciphered with a cipher key: addr_tag = enc[cipher_key](j) || addr_tag_hint
```
```suggestion
/// index ciphered with a cipher key: addr_tag = enc[cipher_key](j)
```

UkoeHB commented on src/seraphis_core/jamtis_support_types.h:
```diff
@@ -57,50 +57,48 @@ struct address_index_t final
+    /// knowledge of the private filter-assist key, whereas AUXILIARY_CHANGE enotes do not. this 1) ensures that if you
+    /// delegate primary view tag scanning to a third party, you will always scan these enotes and 2) prevents
+    /// statistical fingerprints of a third-party scanner seeing multiple primary view tag matches in a single
+    /// transaction. The reason that there is a distinction between EXCLUSIVE_SELF_SPEND and EXCLUSIVE_CHANGE is for UX.
```
```suggestion
    /// In every outgoing transaction that contains owned self-send enotes, there should be *exactly* ONE enote either
    /// with type EXCLUSIVE_SELF_SPEND or EXCLUSIVE_CHANGE, but not both. Any other self-send enotes in that transaction
    /// should have type AUXILIARY_*. Enotes of type EXCLUSIVE_* have primary view tags which are scannable with
    /// knowledge of the private filter-assist key, whereas AUXILIARY_* enotes do not. This 1) ensures that if you
    /// delegate primary view tag scanning to a third party, you will always scan these enotes and 2) prevents
    /// statistical fingerprints of a third-party scanner seeing multiple primary view tag matches in a single
    /// transaction. The reason there is a distinction between *_SELF_SPEND and *_CHANGE is for UX.
```

UkoeHB commented on src/seraphis_impl/serialization_demo_types.h:
```diff
@@ -344,14 +350,17 @@ struct ser_SpMembershipProofV1_PARTIAL final
 struct ser_SpTxSupplementV1 final
 {
-    /// xKe: enote ephemeral pubkeys for outputs
+    /// D_e[]: enote ephemeral pubkeys for outputs
```
```suggestion
    /// D_e: enote ephemeral pubkeys for outputs
```

UkoeHB commented on src/seraphis_impl/serialization_demo_types.h:
```diff
@@ -344,14 +350,17 @@ struct ser_SpMembershipProofV1_PARTIAL final
     std::vector<crypto::x25519_pubkey> output_enote_ephemeral_pubkeys;
     /// tx memo
     std::vector<unsigned char> tx_extra;
+    /// npbits
+    std::uint8_t num_primary_view_tag_bits;
```
These are being serialized, not inferred? What is the consensus model for npbits?

jeffro256 replied:
Yes, since I think it would be a good idea to be able to adjust `npbits` according to relay rules. Otherwise, it would have to be inferred from some version number, and updating that version number would result in a hard fork.

UkoeHB replied:
Ok

UkoeHB commented on src/seraphis_impl/serialization_demo_utils.cpp:
```diff
@@ -411,9 +416,10 @@ void make_serializable_sp_tx_squashed_v1(const SpTxSquashedV1 &tx, ser_SpTxSquas
+    serializable_dest_out.addr_Ks = dest.addr_Ks;
+    serializable_dest_out.addr_Dfa = dest.addr_Dfa;
+    serializable_dest_out.addr_Dvr = dest.addr_Dvr;
+    serializable_dest_out.addr_Dbase = dest.addr_Dbase;
```
```suggestion
    serializable_dest_out.addr_Ks    = dest.addr_Ks;
    serializable_dest_out.addr_Dfa   = dest.addr_Dfa;
    serializable_dest_out.addr_Dvr   = dest.addr_Dvr;
    serializable_dest_out.addr_Dbase = dest.addr_Dbase;
```

UkoeHB commented on src/seraphis_impl/serialization_demo_utils.cpp:
```diff
@@ -729,9 +740,10 @@ bool try_recover_sp_tx_squashed_v1(ser_SpTxSquashedV1 &serializable_tx_in, SpTxS
+    dest_out.addr_Ks = serializable_destination.addr_Ks;
+    dest_out.addr_Dfa = serializable_destination.addr_Dfa;
+    dest_out.addr_Dvr = serializable_destination.addr_Dvr;
+    dest_out.addr_Dbase = serializable_destination.addr_Dbase;
```
```suggestion
    dest_out.addr_Ks    = serializable_destination.addr_Ks;
    dest_out.addr_Dfa   = serializable_destination.addr_Dfa;
    dest_out.addr_Dvr   = serializable_destination.addr_Dvr;
    dest_out.addr_Dbase = serializable_destination.addr_Dbase;
```

UkoeHB commented on src/seraphis_main/enote_record_types.h:
```diff
@@ -129,32 +129,28 @@ struct LegacyEnoteRecord final
 // SpIntermediateEnoteRecordV1 (jamtis non-selfsend enote type only)
-// - a seraphis enote with info extracted using a jamtis find-received key, generate-address secret, and unlock-amounts key
+// - a seraphis enote with info extracted using a jamtis view-received key and generate-address secret key
 ///
-struct SpIntermediateEnoteRecordV1 final
+struct SpIntermediateEnoteRecordV1: public SpBasicEnoteRecordV1
```
I strongly dislike structural inheritance, please find a way to revert this.

jeffro256 replied:
Would it be better to nest the structs such that intermediate enote record has a member field (e.g. `basic`), and the full enote record has an intermediate record as a member field (e.g `intermediate`)?

UkoeHB replied:
Intermediate and basic records are separate things, I'm not sure why they need to be combined. This is basically a refactor anyway, which should be in a separate PR.

jeffro256 replied:
IIRC, writing the structs like this made the enote record utils code more compact, which makes sense since the information in each is a superset of the previous. I will revert tho

jeffro256 replied:
Okay done

UkoeHB replied:
>  writing the structs like this made the enote record utils code more compact

Sometimes what's needed for code clarity is more LoC, not less. Structural inheritance means deeper data access (law of Demeter) and non-local reasoning about a struct's contents.

UkoeHB reviewed (changes requested):
Part 2 (20/72 files done)

jeffro256 commented:
Rebased

UkoeHB committed:
blake2b whitespace fixes

Commit:
SpTxCoinbaseV1: remove block_reward field (#38)

* SpTxCoinbaseV1: remove block_reward field

Not storing/serializing `block_reward` saves us a few bytes on coinbase transactions, and makes it so that you can't
initialize a coinbase transaction that has a block reward not matching its output sum.

jeffro256 commented:
Rebased again

Commit:
seraphis_crypto: split BulletproofPlus2 into proof-only and complete structs (#31)

Commit:
Remaining tx ridigity issues (#34)


---------

Co-authored-by: UkoeHB <37489173+UkoeHB@users.noreply.github.com>

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -160,27 +154,33 @@ void get_coinbase_output_proposal_v1(const JamtisPaymentProposalV1 &proposal,
+    
+    // 8. make standard view tag
+    jamtis::make_jamtis_standard_view_tag(dhe_fa, output_enote_core_out.onetime_address, q,
+        proposal.num_primary_view_tag_bits, view_tag_out);
```
```suggestion
    jamtis::make_jamtis_standard_view_tag(dhe_fa,
        output_enote_core_out.onetime_address,
        q,
        proposal.num_primary_view_tag_bits,
        view_tag_out);
```

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -200,38 +200,44 @@ void get_output_proposal_v1(const JamtisPaymentProposalV1 &proposal,
+    
+    // 9. make standard view tag
+    jamtis::make_jamtis_standard_view_tag(dhe_fa, output_proposal_core_out.onetime_address, q,
+        proposal.num_primary_view_tag_bits, view_tag_out);
```
```suggestion
    jamtis::make_jamtis_standard_view_tag(dhe_fa,
        output_proposal_core_out.onetime_address,
        q,
        proposal.num_primary_view_tag_bits,
        view_tag_out);
```

UkoeHB commented on src/seraphis_core/jamtis_payment_proposal.cpp:
```diff
@@ -286,26 +292,37 @@ void get_output_proposal_v1(const JamtisPaymentProposalSelfSendV1 &proposal,
+    else // exclusive (AKA standard) view tag
+    {
+        jamtis::make_jamtis_standard_view_tag(dhe_fa, output_proposal_core_out.onetime_address, q,
+            proposal.num_primary_view_tag_bits, view_tag_out);
```
```suggestion
        jamtis::make_jamtis_standard_view_tag(dhe_fa,
            output_proposal_core_out.onetime_address,
            q,
            proposal.num_primary_view_tag_bits,
            view_tag_out);
```

UkoeHB commented on src/seraphis_main/tx_builders_outputs.h:
```diff
@@ -159,14 +157,9 @@ void make_v1_outputs_v1(const std::vector<SpOutputProposalV1> &output_proposals,
-*     - A self-send dummy will only be made if there are no other self-sends; otherwise dummies will be purely random.
+*   - add an exclusive/auxiliary change output if necessary
+*   - All output sets will contain exactly 1 exclusive self-send, plus any number of auxiliary self-sends, either from
+*     the original set passed in, or by adding a exclusive/auxiliary change enote here.
```
```suggestion
*     the original set passed in, or by adding an exclusive/auxiliary change enote here.
```

UkoeHB commented on src/seraphis_main/enote_record_types.h:
```diff
@@ -129,30 +129,34 @@ struct LegacyEnoteRecord final
+// - a seraphis enote that resides in a transaction with an enote that passes the primary view tag check
 ///
-struct SpBasicEnoteRecordV1 final
+struct SpBasicEnoteRecordV1
```
Return `final`

UkoeHB commented on src/seraphis_main/enote_record_types.h:
```diff
@@ -129,30 +129,34 @@ struct LegacyEnoteRecord final
+// - a seraphis enote with info extracted using a jamtis view-received key and generate-address secret key
 ///
-struct SpIntermediateEnoteRecordV1 final
+struct SpIntermediateEnoteRecordV1
```
Return `final`

UkoeHB commented on src/seraphis_main/enote_record_types.h:
```diff
@@ -167,28 +171,30 @@ struct SpIntermediateEnoteRecordV1 final
 // - a seraphis enote that has been fully view-scanned with a jamtis view-balance key
 ///
-struct SpEnoteRecordV1 final
+struct SpEnoteRecordV1
```
Return `final`

UkoeHB commented on src/seraphis_main/enote_record_types.h:
```diff
@@ -129,30 +129,34 @@ struct LegacyEnoteRecord final
-    /// t'_addr: nominal address tag (only useful for jamtis non-selfsend enote types)
-    jamtis::address_tag_t nominal_address_tag;
+    /// did this enote not pass the check?
+    bool is_auxiliary;
```
```suggestion
    /// indicates if this enote passed the exclusive-enote view tag check
    /// - If it did not pass, then it *might* be an auxiliary enote.
    bool passed_exclusive_check;
```
The `is_auxiliary` you have is ambiguous.

UkoeHB commented on src/seraphis_main/sp_knowledge_proof_utils.cpp:
```diff
@@ -310,7 +312,7 @@ void make_enote_ownership_proof_v1(const rct::key &jamtis_address_spend_key,
+            .addr_Ks = jamtis_address_spend_key,
             .q   = sender_receiver_secret,
             .C   = amount_commitment,
             .Ko  = onetime_address
```
```suggestion
            .addr_Ks = jamtis_address_spend_key,
            .q       = sender_receiver_secret,
            .C       = amount_commitment,
            .Ko      = onetime_address
```

UkoeHB commented on src/seraphis_main/sp_knowledge_proof_utils.cpp:
```diff
@@ -387,12 +389,14 @@ void make_enote_ownership_proof_v1_receiver(const SpEnoteRecordV1 &enote_record,
-    jamtis::make_jamtis_generateaddress_secret(k_view_balance, s_generate_address);
+    jamtis::make_jamtis_viewreceived_key(k_view_balance, d_view_received);
+    jamtis::make_jamtis_filterassist_key(d_view_received, d_filter_assist);
+    jamtis::make_jamtis_generateaddress_secret(d_view_received, s_generate_address);
```
`d_fa` is not used in this proof.

UkoeHB commented on src/seraphis_main/tx_builders_mixed.cpp:
```diff
@@ -419,10 +419,14 @@ static void check_tx_proposal_semantics_selfsend_outputs_v1(const std::size_t nu
     {
         check_jamtis_payment_proposal_selfsend_semantics_v1(selfsend_payment_proposal,
+
             input_context,
```
```suggestion
        check_jamtis_payment_proposal_selfsend_semantics_v1(selfsend_payment_proposal,
            input_context,
```

UkoeHB commented on src/seraphis_main/tx_component_types.cpp:
```diff
@@ -56,7 +56,7 @@ void append_to_transcript(const SpCoinbaseEnoteV1 &container, SpTranscriptBuilde
     transcript_inout.append("core", container.core);
     transcript_inout.append("addr_tag_enc", container.addr_tag_enc.bytes);
-    transcript_inout.append("view_tag", container.view_tag);
+    transcript_inout.append("double_edged_view_tag", container.view_tag.bytes);
```
```suggestion
    transcript_inout.append("view_tag", container.view_tag.bytes);
```
Keep tags short and concise where possible.

UkoeHB commented on src/seraphis_main/tx_component_types.cpp:
```diff
@@ -71,7 +71,7 @@ void append_to_transcript(const SpEnoteV1 &container, SpTranscriptBuilder &trans
     transcript_inout.append("encoded_amount", container.encoded_amount.bytes);
     transcript_inout.append("addr_tag_enc", container.addr_tag_enc.bytes);
-    transcript_inout.append("view_tag", container.view_tag);
+    transcript_inout.append("double_edged_view_tag", container.view_tag.bytes);
```
```suggestion
    transcript_inout.append("view_tag", container.view_tag.bytes);
```

UkoeHB commented on src/seraphis_main/tx_validators.h:
```diff
@@ -216,6 +219,7 @@ bool validate_sp_semantics_layout_v1(const std::vector<LegacyRingSignatureV4> &l
     const std::vector<SpEnoteV1> &outputs,
     const std::vector<crypto::x25519_pubkey> &enote_ephemeral_pubkeys,
+     const std::uint8_t num_primary_view_tag_bits,
     const TxExtra &tx_extra);
```
```suggestion
    const std::vector<crypto::x25519_pubkey> &enote_ephemeral_pubkeys,
    const std::uint8_t num_primary_view_tag_bits,
    const TxExtra &tx_extra);
```

UkoeHB commented on src/seraphis_mocks/make_mock_tx.h:
```diff
@@ -86,6 +86,7 @@ struct SpTxParamPackV1
     std::vector<rct::xmr_amount> sp_input_amounts{};
     std::vector<rct::xmr_amount> output_amounts{};
     DiscretizedFee discretized_fee{sp::discretize_fee(0)};
+    std::uint8_t num_primary_view_tag_bits{8};
```
```suggestion
    std::uint8_t num_primary_view_tag_bits{0};
```
The param pack default-initializes everything to zero.

UkoeHB reviewed (changes requested):
Part 3 (44/72 files done)

Skipping over the files with huge diffs for now.

DangerousFreedom1984 commented:
Wouldn't it be better to use
1) `sr1` instead of `q`
2) `sr2` instead of `baked_key`  ?

DangerousFreedom1984 commented:
Okay, I reviewed the new changes and compared to the new seraphis draft paper and it is looking good to me. Very nice work!

jeffro256 commented:
@DangerousFreedom1984 It might be better to use that notation in order to match the implementation paper. It also might be a little easier to read them wrong when skimming. I don't have a super strong opinion either way.

UkoeHB committed:
cleanup

UkoeHB committed:
cleanup

Commit:
Direct & compact tx serialization  (#39)

* direct & compact tx serialization

txs are [de]serialized directly from their classes and sizes of containers are not serialized
if they can be implied.

j-berman committed:
Async wallet scanner (#23)

Implement async wallet scanner.

Adds a new functional test for direct wallet2 -> live RPC daemon
interactions. This sets up a framework to test pointing the
Seraphis wallet lib to a live daemon.

Tests sending and scanning:
- a normal transfer
- a sweep single (0 change)
- to a single subaddress
- to 3 subaddresses (i.e. scanning using additional pub keys)

* scan machine: option to force reorg avoidance increment first pass

- when pointing to a daemon that does not support returning empty
blocks when the client requests too high of a height, we have to
be careful in our scanner to always request blocks below the chain
tip, in every request.
- by forcing the reorg avoidance increment on first pass, we make
sure clients will always include the reorg avoidance increment
when requesting blocks from the daemon, so the client can expect
the request for blocks should *always* return an ok height.

* core tests: check conversion tool on all legacy enote version types

Stil TODO:
- check complete scanning on all enote types
- hit every branch condition for all enote versions

* conn pool mock: epee http client connection pool

- Enables concurrent network requests using the epee http client.
- Still TODO for production:

1) close_connections
2) require the pool respect max_connections

* enote finding context: IN LegacyUnscannedChunk, OUT ChunkData

- finds owned enotes by legacy view scanning a chunk of blocks

* async: function to remove minimum element from token queue

- Useful when we want to remove elements of the token queue in an
order that is different than insertion order.

* async scanner: scan via RPC, fetching & scanning parallel chunks

*How it works*

Assume the user's wallet must start scanning blocks from height
5000.

1. The scanner begins by launching 10 RPC requests in parallel to
fetch chunks of blocks as follows:

```
request 0: { start_height: 5000, max_block_count: 20 }
request 1: { start_height: 5020, max_block_count: 20 }
...
request 9: { start_height: 5180, max_block_count: 20 }
```

2. As soon as any single request completes, the wallet immediately
parses that chunk.
  - This is all in parallel. For example, as soon as request 7
  responds, the wallet immediately begins parsing that chunk in
  parallel to any other chunks it's already parsing.
3. If a chunk does not include a total of max_block_count blocks,
and the chunk is not the tip of the chain, this means there was a
"gap" in the chunk request. The scanner launches another parallel
RPC request to fill in the gap.
  - This gap can occur because the server will **not** return a
  chunk of blocks greater than 100mb (or 20k txs) via the
  /getblocks.bin` RPC endpoint
  ([`FIND_BLOCKCHAIN_SUPPLEMENT_MAX_SIZE`](https://github.com/monero-project/monero/blob/053ba2cf07649cea8134f8a188685ab7a5365e5c/src/cryptonote_core/blockchain.cpp#L65))
  - The gap is from `(req.start_height + res.blocks.size())` to
  `(req.start_height + req.max_block_count)`.
4. As soon as the scanner finishes parsing the chunk, it
immediately submits another parallel RPC request.
5. In parallel, the scanner identifies a user's received (and
spent) enotes in each chunk.
  - For example, as soon as request 7 responds and the wallet
  parses it, the wallet scans that chunk in parallel to any other
  chunks it's already scanning.
6. Once a single chunk is fully scanned locally, the scanner
launches a parallel task to fetch and scan the next chunk.
7. Once the scanner reaches the tip of the chain (the terminal
chunk), the scanner terminates.

*Some technical highlights*

- The wallet scanner is backwards compatible with existing daemons
(though it would need to point to an updated daemon to realize the
perf speed-up).
- On error cases such as the daemon going offline, the same wallet
errors that wallet2 uses (that the wallet API expects) are
propagated up to the higher-level Seraphis lib.
- The implementation uses an http client connection pool (reusing
the epee http client) to support parallel network requests
([related](https://github.com/seraphis-migration/wallet3/issues/58)).
- A developer using the scanner can "bring their own blocks/network
implementation" to the scanner by providing a callback function of
the following type as a param to the async scanner constructor:
`std::function<bool(const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::request, cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response)>`

---------

Co-authored-by: jeffro256 <jeffro256@tutanota.com>

Commit:
serialization: serialize tools::variant (#49)

UkoeHB committed:
async: default threadpool cannot be inlined

SNeedlewoods committed:
Identify non-standard received legacy enotes (#43)


---------

Co-authored-by: SNeedlewoods <sneedlewoods_1@protonmail.com>

Commit:
Legacy Indexing Overhaul (#42)

This PR removes "universal"-style indexing for legacy CLSAG rings, and replaces it with a reference set scheme that uses
(amount, index in amount) indexing pairs to reference on-chain enotes. This is the same method that Cryptonote txs use, and is how
the current Monero Core LMDB database is referenced. Doing things this way means that the database will not have to be re-indexed,
saving at a very minimum 1.6 GB (100M on-chain enotes * (16 bytes for extra table keys)) of storage space, and an expensive
database migration involving moving all existing enote data to a new table. We change the MockLedgerContext to support this indexing
scheme.

In practice, serialized txs under this method shouldn't take up much more space than pre-PR if compressed clever-ly, and assuming
most ring members will RingCT enotes.

We also add LegacyEnoteOriginContext for contextualized enote records so we can better keep tracked of scanned legacy enotes under
the legacy indexing scheme.

Co-authored-by: SNeedlewoods <sneedlewoods_1@protonmail.com>

jeffro256 commented:
Rebased to fix conflicts with #23, #39, and #42

UkoeHB commented:
@jeffro256 Sorry the rebase broke this again. Sadly it looks like the work here may end up never being merged (largely my fault for not finishing review earlier this year).
